### PR TITLE
Render Timeline Elements

### PR DESCRIPTION
**Pull request by JamieB-gu**
Implements some of what's needed to support Timeline elements. We're opening this PR as-is because it includes most of the core functionality. Remaining tasks are listed in #10869 and can be completed in follow-up PRs. There aren't any Timeline elements in production yet, so getting this merged before everything is complete shouldn't cause a problem.

Paired with @alinaboghiu for some of these changes.

## What's Included So Far?

1. A `Timeline` component to render both Timeline element variants: flat and sectioned.
2. Stories for the component, showing different features: flat, sectioned, labels, titles, images etc..
3. A change in `renderElement` to handle Timeline elements using this new component.
4. A `NestedArticleElement` wrapper component, to make it easier to render the nested elements that appear in the body of timelines. This could a) perhaps do with a better name, and b) also be used in the existing Key Takeaways and Q And A Explainer components for the same purpose.
5. A change to the timelines enhancer to drop any main media with a role that's not either `inline`, `showcase` or `thumbnail`. Design do not want other roles to be supported.
6. A new `Heading` component, to make code that needs to switch between different levels of headings easier to understand.

### CODE REVIEW

File: dotcom-rendering/src/components/Timeline.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	format,
+	sectioned,
+	smallDates,
+}: EventHeaderProps) => {
+	const heading = (
+		<Heading level={sectioned ? 3 : 2}>
+			<span css={smallDates ? smallDateStyles : titleStyles(format)}>
+				{event.date}
+			</span>
+			{event.title !== undefined ? (
```
Comment (mxdvl) [inline dotcom-rendering/src/components/Timeline.tsx:88]: `hasTitle`?

File: dotcom-rendering/src/components/Timeline.tsx [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	<>
+		{event.label !== undefined && (
+			<div css={labelStyles}>{event.label}</div>
+		)}
+		<section
+			css={[
+				eventStyles,
+				hasShowcaseRole(event.main)
+					? immersiveMainElementEventStyles
+					: undefined,
```
Comment (mxdvl) [inline dotcom-rendering/src/components/Timeline.tsx:170]: Emotion will strip out falsey values if you prefer:

```suggestion
				hasShowcaseRole(event.main)
					&& immersiveMainElementEventStyles
```

File: dotcom-rendering/src/model/enhanceTimeline.ts [hunk 3]
```diff
@@ -11,6 +11,22 @@ import type {
...
+const enhanceMainMedia = (
+	mainMedia: FEElement | undefined,
+): FEElement | undefined =>
+	mainMedia !== undefined &&
+	'role' in mainMedia &&
+	mainMedia.role !== 'showcase' &&
+	mainMedia.role !== 'inline' &&
+	mainMedia.role !== 'thumbnail'
+		? undefined
+		: mainMedia;
```
Comment (mxdvl) [inline dotcom-rendering/src/model/enhanceTimeline.ts:28]: Worth adding a test? Reading this I'm wondering if we're not doing the opposite of the description.

**Review by mxdvl — APPROVED** [review]
Timeline of this review

1. Saw number of added lines
2. Sighed
3. Rolled sleeves up
4. Made tiny comments
5. Approved

**Review by domlander — APPROVED** [review]

File: dotcom-rendering/src/components/Timeline.tsx (hunk 2 above)
Comment (JamieB-gu) [reply dotcom-rendering/src/components/Timeline.tsx:170]: I'm afraid I'm not keen on using implicit conversions and `boolean` operators for values that aren't booleans. I don't find the code very clear, and it can sometimes cause subtle bugs.

File: dotcom-rendering/src/model/enhanceTimeline.test.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
+import assert from 'node:assert/strict';
```
Comment (mxdvl) [inline dotcom-rendering/src/model/enhanceTimeline.test.ts:1]: While I think it’s really nifty to use the [native Node test runner](https://nodejs.org/docs/latest-v20.x/api/test.html), I think it’s best to do this in a separate PR and keep Jest for consistency?

Comment (JamieB-gu) [reply dotcom-rendering/src/model/enhanceTimeline.test.ts:1]: These aren't used for testing our application code, they're included primarily as a form of type guard, for example to prevent TS errors when indexing into arrays. I'm not sure using something like Jest's `expect` will do this (i.e. narrow the types)?

I think `assert` should be preferable to using a type cast, as the error will hopefully be more accurate; type casts often result in unexpected behaviour, or errors downstream that are harder to trace back to their origin.

If there's another way to do this within Jest I'm happy to switch to that!

Comment (mxdvl) [reply dotcom-rendering/src/model/enhanceTimeline.test.ts:1]: That’s a perfectly valid use-case and [definitely superior to casting](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e66570914d31929401f536a7d70a46ef8e290b56/types/node/assert.d.ts#L12)… I wrongly assumed this was related to testing, but it’s [a distinct module](https://nodejs.org/docs/latest-v20.x/api/assert.html).

The [`console.assert()`](https://developer.mozilla.org/en-US/docs/Web/API/console/assert_static) method will not throw an error in the same way and cannot be used instead.
